File: app/src/main/java/com/tunjid/fingergestures/viewmodels/AppViewModel.java
package com.tunjid.fingergestures.viewmodels;

import android.app.Application;
import android.content.pm.ApplicationInfo;

import com.tunjid.fingergestures.App;
import com.tunjid.fingergestures.R;
import com.tunjid.fingergestures.activities.MainActivity;
import com.tunjid.fingergestures.gestureconsumers.GestureMapper;
import com.tunjid.fingergestures.gestureconsumers.RotationGestureConsumer;
import com.tunjid.fingergestures.models.AppState;
import com.tunjid.fingergestures.models.UiState;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.Arrays;
import java.util.Optional;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import androidx.annotation.IntDef;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.recyclerview.widget.DiffUtil;
import io.reactivex.Flowable;
import io.reactivex.Single;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.processors.PublishProcessor;

import static com.tunjid.fingergestures.activities.MainActivity.ACCESSIBILITY_CODE;
import static com.tunjid.fingergestures.activities.MainActivity.DO_NOT_DISTURB_CODE;
import static com.tunjid.fingergestures.activities.MainActivity.SETTINGS_CODE;
import static com.tunjid.fingergestures.activities.MainActivity.STORAGE_CODE;
import static io.reactivex.android.schedulers.AndroidSchedulers.mainThread;

public class AppViewModel extends AndroidViewModel {

    public static final int PADDING = -1;
    public static final int SLIDER_DELTA = PADDING + 1;
    public static final int SLIDER_POSITION = SLIDER_DELTA + 1;
    public static final int SLIDER_DURATION = SLIDER_POSITION + 1;
    public static final int SLIDER_COLOR = SLIDER_DURATION + 1;
    public static final int SCREEN_DIMMER = SLIDER_COLOR + 1;
    public static final int USE_LOGARITHMIC_SCALE = SCREEN_DIMMER + 1;
    public static final int SHOW_SLIDER = USE_LOGARITHMIC_SCALE + 1;
    public static final int ADAPTIVE_BRIGHTNESS = SHOW_SLIDER + 1;
    public static final int ADAPTIVE_BRIGHTNESS_THRESH_SETTINGS = ADAPTIVE_BRIGHTNESS + 1;
    public static final int DOUBLE_SWIPE_SETTINGS = ADAPTIVE_BRIGHTNESS_THRESH_SETTINGS + 1;
    public static final int MAP_UP_ICON = DOUBLE_SWIPE_SETTINGS + 1;
    public static final int MAP_DOWN_ICON = MAP_UP_ICON + 1;
    public static final int MAP_LEFT_ICON = MAP_DOWN_ICON + 1;
    public static final int MAP_RIGHT_ICON = MAP_LEFT_ICON + 1;
    public static final int AD_FREE = MAP_RIGHT_ICON + 1;
    public static final int REVIEW = AD_FREE + 1;
    public static final int WALLPAPER_PICKER = REVIEW + 1;
    public static final int WALLPAPER_TRIGGER = WALLPAPER_PICKER + 1;
    public static final int ROTATION_LOCK = WALLPAPER_TRIGGER + 1;
    public static final int EXCLUDED_ROTATION_LOCK = ROTATION_LOCK + 1;
    public static final int ENABLE_WATCH_WINDOWS = EXCLUDED_ROTATION_LOCK + 1;
    public static final int POPUP_ACTION = ENABLE_WATCH_WINDOWS + 1;
    public static final int ENABLE_ACCESSIBILITY_BUTTON = POPUP_ACTION + 1;
    public static final int ACCESSIBILITY_SINGLE_CLICK = ENABLE_ACCESSIBILITY_BUTTON + 1;
    public static final int ANIMATES_SLIDER = ACCESSIBILITY_SINGLE_CLICK + 1;
    public static final int ANIMATES_POPUP = ANIMATES_SLIDER + 1;
    public static final int DISCRETE_BRIGHTNESS = ANIMATES_POPUP + 1;
    public static final int AUDIO_DELTA = DISCRETE_BRIGHTNESS + 1;
    public static final int AUDIO_STREAM_TYPE = AUDIO_DELTA + 1;
    public static final int AUDIO_SLIDER_SHOW = AUDIO_STREAM_TYPE + 1;
    public static final int NAV_BAR_COLOR = AUDIO_SLIDER_SHOW + 1;
    public static final int LOCKED_CONTENT = NAV_BAR_COLOR + 1;
    public static final int SUPPORT = LOCKED_CONTENT + 1;

    public final int[] gestureItems = {PADDING, MAP_UP_ICON, MAP_DOWN_ICON, MAP_LEFT_ICON,
            MAP_RIGHT_ICON, AD_FREE, SUPPORT, REVIEW, LOCKED_CONTENT, PADDING};

    public final int[] brightnessItems = {PADDING, SLIDER_DELTA, DISCRETE_BRIGHTNESS,
            SCREEN_DIMMER, USE_LOGARITHMIC_SCALE, SHOW_SLIDER, ADAPTIVE_BRIGHTNESS, ANIMATES_SLIDER,
            ADAPTIVE_BRIGHTNESS_THRESH_SETTINGS, DOUBLE_SWIPE_SETTINGS, PADDING};

    public final int[] audioItems = {PADDING, AUDIO_DELTA, AUDIO_STREAM_TYPE, AUDIO_SLIDER_SHOW,
            PADDING};

    public final int[] popupItems = {PADDING, ENABLE_ACCESSIBILITY_BUTTON,
            ACCESSIBILITY_SINGLE_CLICK, ANIMATES_POPUP, ENABLE_WATCH_WINDOWS, ROTATION_LOCK,
            EXCLUDED_ROTATION_LOCK, POPUP_ACTION, PADDING};

    public final int[] appearanceItems = {PADDING, SLIDER_POSITION, SLIDER_DURATION, NAV_BAR_COLOR,
            SLIDER_COLOR, WALLPAPER_PICKER, WALLPAPER_TRIGGER, PADDING};

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({SLIDER_DELTA, SLIDER_POSITION, SLIDER_DURATION, SLIDER_COLOR, SCREEN_DIMMER,
            SHOW_SLIDER, USE_LOGARITHMIC_SCALE, ADAPTIVE_BRIGHTNESS, ADAPTIVE_BRIGHTNESS_THRESH_SETTINGS,
            DOUBLE_SWIPE_SETTINGS, MAP_UP_ICON, MAP_DOWN_ICON, MAP_LEFT_ICON, MAP_RIGHT_ICON,
            AD_FREE, REVIEW, WALLPAPER_PICKER, WALLPAPER_TRIGGER, ROTATION_LOCK,
            EXCLUDED_ROTATION_LOCK, ENABLE_WATCH_WINDOWS, POPUP_ACTION, ENABLE_ACCESSIBILITY_BUTTON,
            ACCESSIBILITY_SINGLE_CLICK, ANIMATES_SLIDER, ANIMATES_POPUP, DISCRETE_BRIGHTNESS,
            AUDIO_DELTA, AUDIO_STREAM_TYPE, AUDIO_SLIDER_SHOW, NAV_BAR_COLOR, LOCKED_CONTENT, SUPPORT})
    public @interface AdapterIndex {}

    public final AppState state;
    private UiState uiState;
    private final String[] quips;
    private final AtomicInteger quipCounter;
    private final CompositeDisposable disposable;
    private final PublishProcessor<UiState> stateProcessor;
    private final PublishProcessor<String> shillProcessor;

    public AppViewModel(@NonNull Application application) {
        super(application);

        uiState = new UiState(application);
        state = new AppState(application);
        disposable = new CompositeDisposable();
        quipCounter = new AtomicInteger(-1);

        stateProcessor = PublishProcessor.create();
        shillProcessor = PublishProcessor.create();

        quips = application.getResources().getStringArray(R.array.upsell_text);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        disposable.clear();
        state.permissionsQueue.clear();
    }

    public Flowable<UiState> uiState() { return stateProcessor.distinct(); }

    public Flowable<String> shill() {
        disposable.clear();
        startShilling();
        return shillProcessor;
    }

    public Single<DiffUtil.DiffResult> updatedApps() {
        return App.diff(state.installedApps,
                () -> getApplication().getPackageManager().getInstalledApplications(0).stream()
                        .filter(this::isUserInstalledApp)
                        .sorted(RotationGestureConsumer.getInstance().getApplicationInfoComparator())
                        .collect(Collectors.toList()),
                info -> info.packageName);
    }

    public Single<DiffUtil.DiffResult> updatedActions() {
        return App.diff(state.availableActions, () -> IntStream.of(GestureMapper.getInstance().getActions()).boxed().collect(Collectors.toList()));
    }

    public void shillMoar() {
        shillProcessor.onNext(getNextQuip());
    }

    public void calmIt() {
        disposable.clear();
    }

    public void checkPermissions() {
        if (state.permissionsQueue.isEmpty())
            stateProcessor.onNext(uiState = uiState.visibility(false));
        else onPermissionAdded();
    }

    public void requestPermission(@MainActivity.PermissionRequest int permission) {
        if (!state.permissionsQueue.contains(permission)) state.permissionsQueue.add(permission);
        onPermissionAdded();
    }

    public void onPermissionClicked(Consumer<Integer> consumer) {
        if (state.permissionsQueue.isEmpty()) checkPermissions();
        else consumer.accept(state.permissionsQueue.peek());
    }

    public Optional<Integer> onPermissionChange(int requestCode) {
        boolean shouldRemove;
        Optional<Integer> result;
        switch (requestCode) {
            case STORAGE_CODE:
                result = Optional.of((shouldRemove = App.hasStoragePermission())
                        ? R.string.storage_permission_granted
                        : R.string.storage_permission_denied);
                break;
            case SETTINGS_CODE:
                result = Optional.of((shouldRemove = App.canWriteToSettings())
                        ? R.string.settings_permission_granted
                        : R.string.settings_permission_denied);
                break;
            case ACCESSIBILITY_CODE:
                result = Optional.of((shouldRemove = App.accessibilityServiceEnabled())
                        ? R.string.accessibility_permission_granted
                        : R.string.accessibility_permission_denied);
                break;
            case DO_NOT_DISTURB_CODE:
                result = Optional.of((shouldRemove = App.hasDoNotDisturbAccess())
                        ? R.string.do_not_disturb_permission_granted
                        : R.string.do_not_disturb_permission_denied);
                break;
            default:
                return Optional.empty();
        }
        if (shouldRemove) state.permissionsQueue.remove(requestCode);
        checkPermissions();
        return result;
    }

    public Optional<Integer> updateBottomNav(int hash) {
        state.permissionsQueue.clear();
        int id = hash == Arrays.hashCode(gestureItems)
                ? R.id.action_directions
                : hash == Arrays.hashCode(brightnessItems)
                ? R.id.action_slider
                : hash == Arrays.hashCode(audioItems)
                ? R.id.action_audio
                : hash == Arrays.hashCode(popupItems)
                ? R.id.action_accessibility_popup
                : hash == Arrays.hashCode(appearanceItems)
                ? R.id.action_wallpaper
                : 0;

        return id == 0 ? Optional.empty() : Optional.of(id);
    }

    private void onPermissionAdded() {
        if (state.permissionsQueue.isEmpty()) return;
        int permissionRequest = state.permissionsQueue.peek();

        switch (permissionRequest) {
            default:
                return;
            case DO_NOT_DISTURB_CODE:
                uiState = uiState.glyph(R.string.enable_do_not_disturb, R.drawable.ic_volume_loud_24dp);
                break;
            case ACCESSIBILITY_CODE:
                uiState = uiState.glyph(R.string.enable_accessibility, R.drawable.ic_human_24dp);
                break;
            case SETTINGS_CODE:
                uiState = uiState.glyph(R.string.enable_write_settings, R.drawable.ic_settings_white_24dp);
                break;
            case STORAGE_CODE:
                uiState = uiState.glyph(R.string.enable_storage_settings, R.drawable.ic_storage_24dp);
                break;
        }

        uiState = uiState.visibility(true);
        stateProcessor.onNext(uiState);
    }

    private void startShilling() {
        disposable.add(Flowable.interval(10, TimeUnit.SECONDS)
                .map(__ -> getNextQuip())
                .observeOn(mainThread())
                .subscribe(shillProcessor::onNext, Throwable::printStackTrace));
    }

    private String getNextQuip() {
        if (quipCounter.incrementAndGet() >= quips.length) quipCounter.set(0);
        return quips[quipCounter.get()];
    }

    private boolean isUserInstalledApp(ApplicationInfo info) {
        return (info.flags & ApplicationInfo.FLAG_UPDATED_SYSTEM_APP) != 0 || (info.flags & ApplicationInfo.FLAG_SYSTEM) == 0;
    }
}

File: app/src/main/java/com/tunjid/fingergestures/viewmodels/AppViewModel.kt
/*
 * Copyright (c) 2017, 2018, 2019 Adetunji Dahunsi.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.tunjid.fingergestures.viewmodels

import android.content.Context
import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.BillingClientStateListener
import com.tunjid.fingergestures.*
import com.tunjid.fingergestures.billing.PurchasesManager
import com.tunjid.fingergestures.di.AppContext
import com.tunjid.fingergestures.di.AppDependencies
import com.tunjid.fingergestures.gestureconsumers.*
import com.tunjid.fingergestures.ui.main.Inputs
import com.tunjid.fingergestures.models.*
import io.reactivex.Flowable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.processors.PublishProcessor
import io.reactivex.rxkotlin.Flowables
import io.reactivex.rxkotlin.addTo
import java.util.*
import java.util.concurrent.TimeUnit
import javax.inject.Inject

class AppViewModel @Inject constructor(
    @AppContext private val context: Context,
    broadcasts: Flowable<Broadcast>,
    override val dependencies: AppDependencies
) : ViewModel(), Inputs {

    private val quips = context.resources.getStringArray(R.array.upsell_text)
    private val inputProcessor: PublishProcessor<Input> = PublishProcessor.create()
    private val disposable: CompositeDisposable = CompositeDisposable()

    private val backingState = Flowables.combineLatest(
        dependencies.purchasesManager.shill(),
        dependencies.purchasesManager.state,
        Flowable.just(context.links),
        broadcasts.filterIsInstance<Broadcast.Prompt>().map { Optional.of(it) }.startWith(Optional.empty()),
        inputProcessor.filterIsInstance<Input.UiInteraction>().startWith(Input.UiInteraction.Default),
        inputProcessor.permissionState,
        inputProcessor.billingState,
        items,
        ::AppState
    )

    val state: LiveData<AppState> = backingState.toLiveData()

    override fun onCleared() {
        disposable.clear()
        state.value?.billingState?.client?.endConnection()
    }

    init {
        val client = BillingClient.newBuilder(context)
            .setListener(dependencies.purchasesManager)
            .build()

        client.startConnection(object : BillingClientStateListener {
            override fun onBillingSetupFinished(@BillingClient.BillingResponse billingResponseCode: Int) =
                accept(Input.Billing.Client(client = client.takeIf { billingResponseCode == BillingClient.BillingResponse.OK }))

            override fun onBillingServiceDisconnected() = accept(Input.Billing.Client(client = null))
        })

        backingState
            .map(AppState::billingState)
            .mapNotNull(BillingState::client)
            .distinctUntilChanged()
            .subscribe {
                val result = it.queryPurchases(BillingClient.SkuType.INAPP)
                if (result.responseCode != BillingClient.BillingResponse.OK) return@subscribe
                dependencies.purchasesManager.onPurchasesQueried(result.responseCode, result.purchasesList)
            }
            .addTo(disposable)

        inputProcessor
            .filterIsInstance<Input.StartTrial>()
            .subscribe { dependencies.purchasesManager.startTrial() }
            .addTo(disposable)
    }

    override fun accept(input: Input): Unit = inputProcessor.onNext(input)

    private fun PurchasesManager.shill() = state
        .map(PurchasesManager.State::hasAds)
        .distinctUntilChanged()
        .switchMap {
            if (it) Flowable.merge(
                inputProcessor.filterIsInstance<Input.Shill>(),
                Flowable.interval(10, TimeUnit.SECONDS)
            ).scan(0) { index, _ -> (index + 1) % quips.size }
                .map(quips::get)
                .map(Shilling::Quip)
            else Flowable.just(Shilling.Calm)
        }

    private val Flowable<Input>.permissionState
        get() = filterIsInstance<Input.Permission>()
            .scan(PermissionState()) { state, permission ->
                when (permission) {
                    Input.Permission.Request.Storage,
                    Input.Permission.Request.Settings,
                    Input.Permission.Request.Accessibility,
                    Input.Permission.Request.DoNotDisturb -> {
                        if (permission is Input.Permission.Request) {
                            val queue = if (state.queue.contains(permission)) state.queue - permission else state.queue
                            state.copy(queue = queue + permission)
                        } else state
                    }
                    is Input.Permission.Action.Clear -> state.copy(
                        queue = listOf()
                    )
                    is Input.Permission.Action.Clicked -> state.copy(
                        active = state.queue.lastOrNull()?.let(::Unique),
                        queue = state.queue.dropLast(1)
                    )
                    is Input.Permission.Action.Changed -> {
                        val (prompt, shouldRemove) = when (permission.request) {
                            Input.Permission.Request.Storage ->
                                if (context.hasStoragePermission) R.string.storage_permission_granted to true
                                else R.string.storage_permission_denied to false
                            Input.Permission.Request.Settings ->
                                if (context.canWriteToSettings) R.string.settings_permission_granted to true
                                else R.string.settings_permission_denied to false
                            Input.Permission.Request.Accessibility ->
                                if (context.accessibilityServiceEnabled) R.string.accessibility_permission_granted to true
                                else R.string.accessibility_permission_denied to false
                            Input.Permission.Request.DoNotDisturb ->
                                if (context.hasDoNotDisturbAccess) R.string.do_not_disturb_permission_granted to true
                                else R.string.do_not_disturb_permission_denied to false
                        }

                        state.copy(
                            prompt = Unique(prompt),
                            queue = when {
                                shouldRemove -> state.queue - permission.request
                                else -> state.queue
                            }
                        )
                    }
                }
            }

//    private fun consume(purchaseToken: String) {
//        billingClient.consumeAsync(purchaseToken) { _, _ -> }
//    }
//
//    private fun consumeAll() {
//        PurchasesManager.instance.clearPurchases()
//        val result = billingClient.queryPurchases(BillingClient.SkuType.INAPP)
//        if (billingClient == null || result.responseCode != BillingClient.BillingResponse.OK) return
//        for (item in result.purchasesList) consume(item.purchaseToken)
//    }

    companion object {
        const val PADDING = -1
        const val SLIDER_DELTA = PADDING + 1
        const val SLIDER_POSITION = SLIDER_DELTA + 1
        const val SLIDER_DURATION = SLIDER_POSITION + 1
        const val SLIDER_COLOR = SLIDER_DURATION + 1
        const val SCREEN_DIMMER = SLIDER_COLOR + 1
        const val USE_LOGARITHMIC_SCALE = SCREEN_DIMMER + 1
        const val SHOW_SLIDER = USE_LOGARITHMIC_SCALE + 1
        const val ADAPTIVE_BRIGHTNESS = SHOW_SLIDER + 1
        const val ADAPTIVE_BRIGHTNESS_THRESH_SETTINGS = ADAPTIVE_BRIGHTNESS + 1
        const val DOUBLE_SWIPE_SETTINGS = ADAPTIVE_BRIGHTNESS_THRESH_SETTINGS + 1
        const val MAP_UP_ICON = DOUBLE_SWIPE_SETTINGS + 1
        const val MAP_DOWN_ICON = MAP_UP_ICON + 1
        const val MAP_LEFT_ICON = MAP_DOWN_ICON + 1
        const val MAP_RIGHT_ICON = MAP_LEFT_ICON + 1
        const val AD_FREE = MAP_RIGHT_ICON + 1
        const val REVIEW = AD_FREE + 1
        const val WALLPAPER_PICKER = REVIEW + 1
        const val WALLPAPER_TRIGGER = WALLPAPER_PICKER + 1
        const val ROTATION_LOCK = WALLPAPER_TRIGGER + 1
        const val EXCLUDED_ROTATION_LOCK = ROTATION_LOCK + 1
        const val ENABLE_WATCH_WINDOWS = EXCLUDED_ROTATION_LOCK + 1
        const val POPUP_ACTION = ENABLE_WATCH_WINDOWS + 1
        const val ENABLE_ACCESSIBILITY_BUTTON = POPUP_ACTION + 1
        const val ACCESSIBILITY_SINGLE_CLICK = ENABLE_ACCESSIBILITY_BUTTON + 1
        const val ANIMATES_SLIDER = ACCESSIBILITY_SINGLE_CLICK + 1
        const val ANIMATES_POPUP = ANIMATES_SLIDER + 1
        const val DISCRETE_BRIGHTNESS = ANIMATES_POPUP + 1
        const val AUDIO_DELTA = DISCRETE_BRIGHTNESS + 1
        const val AUDIO_STREAM_TYPE = AUDIO_DELTA + 1
        const val AUDIO_SLIDER_SHOW = AUDIO_STREAM_TYPE + 1
        const val NAV_BAR_COLOR = AUDIO_SLIDER_SHOW + 1
        const val LOCKED_CONTENT = NAV_BAR_COLOR + 1
        const val SUPPORT = LOCKED_CONTENT + 1
        const val ROTATION_HISTORY = SUPPORT + 1

        internal const val RX_JAVA_LINK = "https://github.com/ReactiveX/RxJava"
        internal const val COLOR_PICKER_LINK = "https://github.com/QuadFlask/colorpicker"
        internal const val ANDROID_BOOTSTRAP_LINK = "https://github.com/tunjid/android-bootstrap"
        internal const val GET_SET_ICON_LINK = "http://www.myiconfinder.com/getseticons"
        internal const val IMAGE_CROPPER_LINK = "https://github.com/ArthurHub/Android-Image-Cropper"
        internal const val MATERIAL_DESIGN_ICONS_LINK = "https://materialdesignicons.com/"
    }
}

val Context.links
    get() = listOf(
        TextLink(getString(R.string.get_set_icon), AppViewModel.GET_SET_ICON_LINK),
        TextLink(getString(R.string.rxjava), AppViewModel.RX_JAVA_LINK),
        TextLink(getString(R.string.color_picker), AppViewModel.COLOR_PICKER_LINK),
        TextLink(getString(R.string.image_cropper), AppViewModel.IMAGE_CROPPER_LINK),
        TextLink(getString(R.string.material_design_icons), AppViewModel.MATERIAL_DESIGN_ICONS_LINK),
        TextLink(getString(R.string.android_bootstrap), AppViewModel.ANDROID_BOOTSTRAP_LINK)
    )

private val Flowable<Input>.billingState
    get() = filterIsInstance<Input.Billing>()
        .scan(BillingState()) { state, item ->
            when (item) {
                is Input.Billing.Client -> state.copy(client = item.client)
                is Input.Billing.Purchase -> when (val client = state.client) {
                    null -> state.copy(prompt = Unique(R.string.billing_not_connected))
                    else -> state.copy(cart = Unique(client to item.sku))
                }
            }
        }

fun <T, R> Flowable<List<T>>.listMap(mapper: (T) -> R): Flowable<List<R>> =
    map { it.map(mapper) }

fun <T : Any, R> Flowable<T>.mapNotNull(mapper: (T) -> R?): Flowable<R> =
    map { Optional.ofNullable(mapper(it)) }
        .filter(Optional<R>::isPresent)
        .map(Optional<R>::get)

inline fun <reified T> Flowable<*>.filterIsInstance(): Flowable<T> =
    filter { it is T }
        .cast(T::class.java)
